java-checks/src/main/java/org/sonar/java/checks/FunctionalInterfacesSpecialisedCheck.java

```diff
@@ -0,0 +1,257 @@
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import com.google.common.collect.ImmutableList;
+import java.util.List;
+import java.util.Objects;
+import javax.annotation.CheckForNull;
+import org.sonar.check.Rule;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.semantic.Type;
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.ParameterizedTypeTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.TypeArguments;
+import org.sonar.plugins.java.api.tree.TypeTree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+import org.sonar.plugins.java.api.tree.WildcardTree;
+
+@Rule(key = "S4276")
+public class FunctionalInterfacesSpecialisedCheck extends IssuableSubscriptionVisitor {
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return ImmutableList.of(Tree.Kind.CLASS, Tree.Kind.NEW_CLASS, Tree.Kind.VARIABLE);
+  }
+
+  private static final String REPORT_STRING = "Refactor this code to use the more specialised Functional Interface '";
+  private static final String UNARY_OPERATOR = "UnaryOperator<";
+
+  @Override
+  public void visitNode(Tree tree) {
+    String reportString = null;
+    if (tree.is(Tree.Kind.CLASS)) {
+      ClassTree interfaceTree = (ClassTree) tree;
+      interfaceTree.superInterfaces().stream().map(FunctionalInterfacesSpecialisedCheck::matchInterface).filter(Objects::nonNull)
+        .forEach(reportMessage -> report(interfaceTree, reportMessage));
+    } else if (tree.is(Tree.Kind.VARIABLE)) {
+      VariableTree variableTree = (VariableTree) tree;
+      TypeTree interfaceType = variableTree.type();
+      if (variableTree.initializer() != null && variableTree.initializer().is(Tree.Kind.NEW_CLASS, Tree.Kind.LAMBDA_EXPRESSION)) {
+        reportString = matchInterface(interfaceType);
+        if (reportString == null) {
+          return;
+        }
+        report(interfaceType, reportString);
+      }
+
+    }
+  }
+
+  @CheckForNull
+  private static String matchInterface(TypeTree interfaceType) {
+    TypeArguments args = interfaceParameters(interfaceType);
+    if (args == null) {
+      return null;
+    }
+    String reportString = null;
+    switch (interfaceType.symbolType().fullyQualifiedName()) {
```

Comment:
if you need the symbol type, it means the check requires semantic. Please add the `hasSemantic()` control at the beginning of the check, and a proper test.